File: L/LCP40/main.go
package main

import (
	"fmt"
	"sort"
)

func maxmiumScore(cards []int, cnt int) int {
	sort.Ints(cards)
	odd := []int{}
	even := []int{}
	for _, v := range cards {
		if v%2 == 0 {
			even = append(even, v)
		} else {
			odd = append(odd, v)
		}
	}
	for i := len(even) - 2; i >= 0; i-- {
		even[i] = even[i] + even[i+1]
	}
	for i := len(odd) - 2; i >= 0; i-- {
		odd[i] = odd[i] + odd[i+1]
	}

	start := cnt
	if start > len(even) {
		start = len(even)
	}
	max := 0
	for i := len(even) - start; i < len(even); i++ {
		remain := cnt - len(even) + i
		if remain%2 == 1 || remain > len(odd) {
			continue
		}

		v := even[i]
		if remain > 0 {
			v += odd[len(odd)-remain]
		}
		if max < v {
			max = v
		}
	}

	if cnt%2 == 0 && cnt <= len(odd) && max < odd[len(odd)-cnt] {
		max = odd[len(odd)-cnt]
	}

	return max
}

func main() {
	fmt.Println()
}

File: L/LCP40/main.rs
struct Solution {}

impl Solution {
  pub fn maxmium_score(cards: Vec<i32>, cnt: i32) -> i32 {
    let mut cards = cards;
    cards.sort_unstable_by(|a, b| b.cmp(a));
    let mut sum: i32 = 0;
    let mut idx: usize = 0;
    while idx < cnt as usize {
      sum += cards[idx];
      idx += 1;
    }
    if sum % 2 == 0 {
      return sum;
    }
    idx -= 1;
    let mut left = idx;
    let mut right = idx + 1;
    let mut sum2 = sum;
    let mut found1: bool = false;
    let mut found2: bool = false;
    loop {
      if cards[left] % 2 == 1 {
        sum2 -= cards[left];
        found1 = true;
        break;
      }
      if left == 0 {
        break;
      }
      left -= 1;
    }
    while right < cards.len() {
      if cards[right] % 2 == 0 {
        sum2 += cards[right];
        found2 = true;
        break;
      }
      right += 1;
    }
    let mut ans = 0;
    if found1 && found2 {
      ans = sum2;
    };

    left = idx;
    right = idx + 1;
    sum2 = sum;
    found1 = false;
    found2 = false;
    loop {
      if cards[left] % 2 == 0 {
        sum2 -= cards[left];
        found1 = true;
        break;
      }
      if left == 0 {
        break;
      }
      left -= 1;
    }
    while right < cards.len() {
      if cards[right] % 2 == 1 {
        sum2 += cards[right];
        found2 = true;
        break;
      }
      right += 1;
    }
    if found1 && found2 {
      ans = ans.max(sum2);
    }

    ans
  }
}

fn main() {
  // println!("{}", Solution::maxmium_score(vec![1, 2, 8, 9], 3));
  println!("{}", Solution::maxmium_score(vec![3, 3, 1], 1));
}
